libcudacxx/include/cuda/__memory_resource/cuda_async_memory_resource.h

```diff
@@ -0,0 +1,282 @@
+      &__ptr,
+      __bytes,
+      __pool_,
+      ::cudaStream_t{0});
+    return __ptr;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated. Must have been allocated through a call to `allocate`
+   * @param __bytes  The number of bytes that was passed to the `allocate` call that returned \p __ptr.
+   * @param __alignment The alignment that was passed to the `allocate` call that returned \p __ptr.
+   */
+  void deallocate(void* __ptr, const size_t, const size_t __alignment = default_cuda_malloc_alignment)
+  {
+    // We need to ensure that the provided alignment matches the minimal provided alignment
+    _LIBCUDACXX_ASSERT(__is_valid_alignment(__alignment),
+                       "Invalid alignment passed to cuda_memory_resource::deallocate.");
+    _CCCL_ASSERT_CUDA_API(::cudaFreeAsync, "cuda_async_memory_resource::deallocate failed", __ptr, ::cudaStream_t{0});
+    (void) __alignment;
+  }
+
+  /**
+   * @brief Allocate device memory of size at least \p __bytes via cudaMallocFromPoolAsync.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation.
+   * @param __stream Stream on which to perform allocation.
+   * @throws cuda::cuda_error of the returned error code.
+   * @return void* Pointer to the newly allocated memory
+   */
+  _CCCL_NODISCARD void* allocate_async(const size_t __bytes, const size_t __alignment, const ::cuda::stream_ref __stream)
+  {
+    // We need to ensure that the provided alignment matches the minimal provided alignment
+    if (!__is_valid_alignment(__alignment))
+    {
+      _CUDA_VSTD_NOVERSION::__throw_bad_alloc();
+    }
+
+    return allocate_async(__bytes, __stream);
+  }
+
+  /**
+   * @brief Allocate device memory of size at least \p __bytes via cudaMallocFromPoolAsync.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __stream Stream on which to perform allocation.
+   * @throws cuda::cuda_error of the returned error code.
+   * @return void* Pointer to the newly allocated memory
+   */
+  _CCCL_NODISCARD void* allocate_async(const size_t __bytes, const ::cuda::stream_ref __stream)
+  {
+    void* __ptr{nullptr};
+    _CCCL_TRY_CUDA_API(
+      ::cudaMallocFromPoolAsync,
+      "cuda_async_memory_resource::allocate_async failed to allocate with cudaMallocFromPoolAsync",
+      &__ptr,
+      __bytes,
+      __pool_,
+      __stream.get());
+    return __ptr;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated. Must have been allocated through a call to `allocate_async`
+   * @param __bytes The number of bytes that was passed to the `allocate_async` call that returned \p __ptr.
+   * @param __alignment The alignment that was passed to the `allocate_async` call that returned \p __ptr.
+   * @param __stream Stream that was passed to the `allocate_async` call that returned \p __ptr.
```

Comment:
I don't think this is right. The deallocate call should be allowed to come on any stream that has a synchronization relationship with the allocation's stream, not just from the exact same stream.

Reply:
That's a great point. Shame on me for not catching this. This applies to any stream-ordered memory resource.

Reply:
We should probably have documentation of stream-ordered allocation semantics. These could be as simple as a reference to the definitions/documentation in the CUDA docs.

Reply:
Thanks for catching this. I rephrased and added a link to the docs